File: C#/ex11/max_subarray_solution.cs
// Max Subarray (Exercise 11) - Solution
// Find the maximum sum of a contiguous subarray using Kadane's algorithm.

using System;
using System.Collections.Generic;

class Program {
    static void Main() {
        int n = int.Parse(Console.ReadLine()!);
        
        string? line = Console.ReadLine();
        if (line == null) return;
        
        var parts = line.Split(new[]{' ', '\t'}, StringSplitOptions.RemoveEmptyEntries);
        List<int> arr = new List<int>();
        foreach (string part in parts) {
            if (int.TryParse(part, out int val)) {
                arr.Add(val);
            }
        }
        
        long maxSum = arr[0];
        long currentSum = arr[0];
        
        for (int i = 1; i < arr.Count; i++) {
            currentSum = Math.Max(arr[i], currentSum + arr[i]);
            maxSum = Math.Max(maxSum, currentSum);
        }
        
        Console.WriteLine(maxSum);
    }
}

File: C/ex12/max_subarray_solution.c
// Max Subarray (Exercise 12) - Solution
// Find the maximum sum of a contiguous subarray using Kadane's algorithm.

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int main() {
    int n;
    scanf("%d", &n);
    
    int *arr = (int *)malloc(n * sizeof(int));
    if (arr == NULL) {
        return 1;
    }
    
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }
    
    long maxSum = arr[0];
    long currentSum = arr[0];
    
    for (int i = 1; i < n; i++) {
        currentSum = (arr[i] > currentSum + arr[i]) ? arr[i] : currentSum + arr[i];
        if (currentSum > maxSum) {
            maxSum = currentSum;
        }
    }
    
    printf("%ld\n", maxSum);
    
    free(arr);
    return 0;
}
